File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/commons/ScreenBasicValidation.java
package com.kishlaly.ta.analyze.tasks.blocks.commons;

import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import static com.kishlaly.ta.model.BlockResultCodeJava.*;
import static com.kishlaly.ta.utils.QuotesJava.resolveMinBarsCount;

public class ScreenBasicValidation implements CommonBlock {

    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        if (screen.quotes.isEmpty() || screen.quotes.size() < resolveMinBarsCount(screen.timeframe)) {
            LogJava.addDebugLine("There are not enough quotes for " + screen.timeframe.name());
            LogJava.recordCode(NO_DATA_QUOTES, screen);
            return new BlockResultJava(null, NO_DATA_QUOTES);
        }

        List<IndicatorJava> missingData = new ArrayList<>();
        screen.indicators.forEach((indicator, value) -> {
            if (value.isEmpty() || value.size() < resolveMinBarsCount(screen.timeframe)) {
                missingData.add(indicator);
            }
        });
        if (!missingData.isEmpty()) {
            LogJava.recordCode(NO_DATA_INDICATORS, screen);
            LogJava.addDebugLine("No indicator data: " + missingData.stream().map(indicator -> indicator.name()).collect(Collectors.joining(", ")));
            return new BlockResultJava(null, NO_DATA_INDICATORS);
        }

        return new BlockResultJava(null, OK);
    }

}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/commons/ScreenBasicValidation.kt
package com.kishlaly.ta.analyze.tasks.blocks.commons

import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.Quote
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.model.indicators.Indicator
import com.kishlaly.ta.utils.Log
import com.kishlaly.ta.utils.Quotes.Companion.resolveMinBarsCount
import com.kishlaly.ta.utils.isValid

class ScreenBasicValidation : CommonBlock {

    override fun check(screen: SymbolData): BlockResult {
        if (!screen.isValid()) {
            Log.addDebugLine("There are not enough quotes for ${screen.timeframe.name}")
            Log.recordCode(BlockResultCode.NO_DATA_QUOTES, screen)
            return BlockResult(Quote.NaN(), BlockResultCode.NO_DATA_QUOTES)
        }

        val missingData = mutableListOf<Indicator>()
        screen.indicators.forEach { indicator, value ->
            if (value.isEmpty() || value.size < resolveMinBarsCount(screen.timeframe)) {
                missingData.add(indicator)
            }
        }
        if (!missingData.isEmpty()) {
            Log.recordCode(BlockResultCode.NO_DATA_INDICATORS, screen)
            Log.addDebugLine("No indicator data: " + missingData.map { it.name }.joinToString())
            return BlockResult(Quote.NaN(), BlockResultCode.NO_DATA_INDICATORS)
        }

        return BlockResult(Quote.NaN(), BlockResultCode.OK)
    }

}